lua/mason-null-ls/init.lua

```diff
@@ -1,140 +1,135 @@
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').code_actions))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').completion))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').hover))
-	local tools = lookup(removeArrayDuplicates(sources))
-	return tools
+	return _.filter_map(function(pkg_name)
+		return Optional.of_nilable(source_mapping.package_to_null_ls[pkg_name])
+	end, registry.get_installed_package_names())
 end
 
-local do_install = function(p, version, on_close)
-	if version ~= nil then
-		show(string.format('%s: updating to %s', p.name, version))
-	else
-		show(string.format('%s: installing', p.name))
+---@param filetype string | string[]
+local function is_source_in_filetype(filetype)
+	local filetype_mapping = require('mason-null-ls.mappings.filetype')
+
+	local function get_sources_by_filetype(ft)
+		return filetype_mapping[ft] or {}
 	end
-	p:once('install:success', function()
-		show(string.format('%s: successfully installed', p.name))
-	end)
-	p:once('install:failed', function()
-		show_error(string.format('%s: failed to install', p.name))
-	end)
-	p:install({ version = version }):once('closed', vim.schedule_wrap(on_close))
-end
 
-local check_install = function(force_update)
-	SETTINGS.ensure_installed =
-		vim.tbl_deep_extend('force', lookup(SETTINGS.null_ls_sources), SETTINGS.ensure_installed)
+	local source_candidates = _.compose(
+		_.set_of,
+		_.cond({
+			{ _.is('string'), get_sources_by_filetype },
+			{ _.is('table'), _.compose(_.flatten, _.map(get_sources_by_filetype)) },
+			{ _.T, _.always({}) },
+		})
+	)(filetype)
 
-	if SETTINGS.automatic_installation then
-		SETTINGS.ensure_installed = vim.tbl_deep_extend('force', auto_get_packages(), SETTINGS.ensure_installed)
+	---@param source_name string
+	---@return boolean
+	return function(source_name)
+		return source_candidates[source_name]
 	end
+end
 
-	local completed = 0
-	local total = vim.tbl_count(SETTINGS.ensure_installed)
-	local on_close = function()
-		completed = completed + 1
-		if completed >= total then
-			vim.api.nvim_exec_autocmds('User', {
-				pattern = 'MasonNullLsUpdateCompleted',
-				-- 'data' doesn't work with < 0.8.0
-				-- data = { packages = SETTINGS.ensure_installed },
-			})
-		end
+---Get a list of available sources in mason-registry
+---@param filter { filetype: string | string[] }?: (optional) Used to filter the list of source names.
+--- The available keys are
+---   - filetype (string | string[]): Only return sources with matching filetype
+---@return string[]
+function M.get_available_sources(filter)
+	local registry = require('mason-registry')
+	local source_mapping = require('mason-null-ls.mappings.source')
+	local Optional = require('mason-core.optional')
+	filter = filter or {}
+	local predicates = {}
+
+	if filter.filetype then
+		table.insert(predicates, is_source_in_filetype(filter.filetype))
```

Comment:
i think this is the answer to why you returned a function. it would better to do 
```suggestion
		table.insert(predicates, 
		function(source_name)
		     is_source_in_filetype(source_name, filter.filetype)
		end)
```

Reply:
tried it your way
`require('mason-null-ls').get_available_sources({ filetype = 'lua' })`
started returning an empty table when it normally returns `{'selene', 'luacheck', 'stylua'}`

I'll leave such refactoring up to you since this is a matter of taste at this point

Reply:
ehh the function name is misleading `is*` typical denotes boolean being returned. I am okay with keeping as is